src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2427,11 +2427,52 @@ void QgsVertexTool::applyEditsToLayers( QgsVertexTool::VertexEdits &edits )
         id.insert( it2.key() );
         ignoreFeatures.insert( layer, id );
         int avoidIntersectionsReturn = featGeom.avoidIntersections( avoidIntersectionsLayers, ignoreFeatures );
+
         switch ( avoidIntersectionsReturn )
         {
-          case 2:
-            emit messageEmitted( tr( "The operation would change the geometry type." ), Qgis::MessageLevel::Warning );
+          case 2: // Geometry type was changed, let's try our best to make it compatible with the target layer
+          {
+            const QVector<QgsGeometry> newGeoms = featGeom.coerceToType( layer->wkbType() );
+            if ( newGeoms.count() == 1 )
+            {
+              featGeom = newGeoms.at( 0 );
+            }
+            else // handle multi geometries
+            {
+              QgsFeatureList removedFeatures;
+              double largest = 0;
+              QgsFeature originalFeature = layer->getFeature( it2.key() );
+              int largestPartIndex = -1;
+              for ( int i = 0; i < newGeoms.size(); ++i )
+              {
+                QgsGeometry currentPart = newGeoms.at( i );
+                const double currentPartSize = layer->geometryType() == Qgis::GeometryType::Polygon ? currentPart.area() : currentPart.length();
+
+                QgsFeature partFeature( layer->fields() );
+                partFeature.setAttributes( originalFeature.attributes() );
+                partFeature.setGeometry( currentPart );
+                removedFeatures.append( partFeature );
+                if ( currentPartSize > largest )
+                {
+                  featGeom = currentPart;
+                  largestPartIndex = i;
+                  largest = currentPartSize;
+                }
+              }
+              removedFeatures.removeAt( largestPartIndex );
+              QgsMessageBarItem *messageBarItem = QgisApp::instance()->messageBar()->createMessage( tr( "Avoid overlaps" ), tr( "Only the largest of multiple created geometries was preserved." ) );
+              QPushButton *restoreButton = new QPushButton( tr( "Restore others" ) );
+              connect( restoreButton, &QPushButton::clicked, restoreButton, [ = ]
+              {
+                layer->beginEditCommand( tr( "Restored geometry parts removed by avoid overlaps" ) );
+                QgsFeatureList unconstFeatures = removedFeatures;
+                QgisApp::instance()->pasteFeatures( layer, 0, removedFeatures.size(), unconstFeatures );
+              } );
+              messageBarItem->layout()->addWidget( restoreButton );
+              QgisApp::instance()->messageBar()->pushWidget( messageBarItem, Qgis::MessageLevel::Info, 5 );
```

Comment:
Feel free to propose something else, this was chosen based on other messages.

Reply:
15s? or more ( _+ it is not sure it will get translated for the first times people will face it_)?
I may be wrong but I don't think that most (any?) of the other messages require an immediate decision from the user. They are mainly and simply informative.

Reply:
```suggestion
              QgisApp::instance()->messageBar()->pushWidget( messageBarItem, Qgis::MessageLevel::Info, 15 );
```